src/libs/ObjectUtils.ts

```diff
@@ -4,8 +4,9 @@ const shallowCompare = (obj1?: object, obj2?: object) => {
         return true;
     }
     if (obj1 && obj2) {
-        // @ts-expect-error we know that obj1 and obj2 are params of a route.
-        return Object.keys(obj1).length === Object.keys(obj2).length && Object.keys(obj1).every((key) => obj1[key] === obj2[key]);
+        const keys1 = Object.keys(obj1) as Array<keyof typeof obj1>;
+        const keys2 = Object.keys(obj2) as Array<keyof typeof obj2>;
```

Comment:
Are these assertions necessary?

Reply:
Let's improve `shallowCompare`:

```tsx
function shallowCompare(obj1?: Record<string, unknown>, obj2?: Record<string, unknown>): boolean {
    if (!obj1 && !obj2) {
        return true;
    }

    if (obj1 && obj2) {
        const keys1 = Object.keys(obj1);
        const keys2 = Object.keys(obj2);
        return keys1.length === keys2.length && keys1.every((key) => obj1[key] === obj2[key]);
    }

    return false;
}

export default shallowCompare;
```

And then you have to make the following adjustments:

```diff
diff --git a/src/libs/Navigation/AppNavigator/getPartialStateDiff.ts b/src/libs/Navigation/AppNavigator/getPartialStateDiff.ts
index 4c18e161c9..5061c75007 100644
--- a/src/libs/Navigation/AppNavigator/getPartialStateDiff.ts
+++ b/src/libs/Navigation/AppNavigator/getPartialStateDiff.ts
@@ -53,7 +53,7 @@ function getPartialStateDiff(state: State<RootStackParamList>, templateState: St
             (stateTopmostCentralPane &&
                 templateStateTopmostCentralPane &&
                 stateTopmostCentralPane.name !== templateStateTopmostCentralPane.name &&
-                !shallowCompare(stateTopmostCentralPane.params, templateStateTopmostCentralPane.params))
+                !shallowCompare(stateTopmostCentralPane.params as Record<string, unknown> | undefined, templateStateTopmostCentralPane.params as Record<string, unknown> | undefined))
         ) {
             // We need to wrap central pane routes in the central pane navigator.
             diff[NAVIGATORS.CENTRAL_PANE_NAVIGATOR] = templateStateTopmostCentralPane;
@@ -73,7 +73,7 @@ function getPartialStateDiff(state: State<RootStackParamList>, templateState: St
             (stateTopmostFullScreen &&
                 templateStateTopmostFullScreen &&
                 stateTopmostFullScreen.name !== templateStateTopmostFullScreen.name &&
-                !shallowCompare(stateTopmostFullScreen.params, templateStateTopmostFullScreen.params))
+                !shallowCompare(stateTopmostFullScreen.params as Record<string, unknown> | undefined, templateStateTopmostFullScreen.params as Record<string, unknown> | undefined))
         ) {
             diff[NAVIGATORS.FULL_SCREEN_NAVIGATOR] = fullScreenDiff;
         }
diff --git a/src/libs/Navigation/linkTo.ts b/src/libs/Navigation/linkTo.ts
index 863cb102ad..b21a24a76e 100644
--- a/src/libs/Navigation/linkTo.ts
+++ b/src/libs/Navigation/linkTo.ts
@@ -153,7 +153,7 @@ export default function linkTo(navigation: NavigationContainerRef<RootStackParam
         const isTargetNavigatorOnTop = topRouteName === action.payload.name;
 
         const isTargetScreenDifferentThanCurrent = Boolean(topmostCentralPaneRoute && topmostCentralPaneRoute.name !== action.payload.params?.screen);
-        const areParamsDifferent = !shallowCompare(topmostCentralPaneRoute?.params, action.payload.params?.params);
+        const areParamsDifferent = !shallowCompare(topmostCentralPaneRoute?.params as Record<string, unknown> | undefined, action.payload.params?.params);
 
         // In case if type is 'FORCED_UP' we replace current screen with the provided. This means the current screen no longer exists in the stack
         if (type === CONST.NAVIGATION.TYPE.FORCED_UP) {

```